src/Uno.UI/UI/Xaml/Media/Animation/KeyFrameScheduler.cs

```diff
@@ -0,0 +1,259 @@
+﻿#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Windows.System;
+using Uno.Disposables;
+using Uno.Extensions;
+
+namespace Windows.UI.Xaml.Media.Animation
+{
+	internal class KeyFrameScheduler<TValue> : IDisposable
+	{
+		public delegate IDisposable? OnFrame(TValue current, IKeyFrame<TValue> frame, TimeSpan duration);
+
+		private readonly Stopwatch _watch = new Stopwatch();
+		private readonly OnFrame _onFrame;
+		private readonly Action<EndReason> _onCompleted;
+		private readonly TimeSpan? _beginTime;
+		private readonly TimeSpan? _duration;
+		private readonly List<IKeyFrame<TValue>> _frames;
+
+		private DispatcherQueueTimer? _timer;
+		private int _frameId = -1;
+		private SerialDisposable? _frameSubscription; // Lazy init only if needed (not used for the common discrete key frames)
+		private TimeSpan _seekOffset;
+
+		private int _state = States.NotRunning;
+
+		private static class States
+		{
+			public const int NotRunning = 0;
+			public const int Running = 1;
+			public const int Paused = 2;
+			public const int Ended = int.MaxValue; // a.k.a. Stoped / Disposed : This Scheduler can run frames only once!
+		}
+
+		public enum EndReason
+		{
+			/// <summary>
+			/// Reached the last frame and wait for the configured duration
+			/// </summary>
+			EndOfFrames,
+
+			/// <summary>
+			/// Stopped by the user
+			/// </summary>
+			Stopped,
+
+			/// <summary>
+			/// Aborted
+			/// </summary>
+			Disposed
+		}
+
+		public KeyFrameScheduler(
+			TimeSpan? beginTime,
+			TimeSpan? duration,
+			TValue initialValue,
+			IEnumerable<IKeyFrame<TValue>>? frames,
+			OnFrame onFrame,
+			Action<EndReason> onCompleted)
+		{
+			CurrentValue = initialValue;
+
+			frames ??= Enumerable.Empty<IKeyFrame<TValue>>();
+			frames = frames.Trim();
+			frames = duration.HasValue
+				? frames.Where(k => k != null && k.KeyTime.TimeSpan <= duration.Value)
```

Comment:
nit: Can `k` be null here? We do `.Trim()` above.

Reply:
Good catch, I lost it while playing with git, but I've removed the `.Trim()` above to reduce the number of operators (a bit more performant)